pkg/kubelet/kubelet.go

```diff
@@ -1025,137 +1014,204 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
+				// look for changes in the container.
+				containerChanged := hash != 0 && hash != expectedHash
+				if !containerChanged {
+					// TODO: This should probably be separated out into a separate goroutine.
+					// If the container's liveness probe is unsuccessful, set readiness to false. If liveness is succesful, do a readiness check and set
+					// readiness accordingly. If the initalDelay since container creation on liveness probe has not passed the probe will return Success.
+					// If the initial delay on the readiness probe has not passed the probe will return Failure.
+					ready := probe.Unknown
+					live, err := kl.probeContainer(container.LivenessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Success)
+					if err != nil {
+						glog.Infof("liveness/readiness probe errored: %v", err)
+						containersToKeep[containerID] = &container
+						continue
+					}
+					if live == probe.Success {
+						ready, _ = kl.probeContainer(container.ReadinessProbe, podFullName, uid, podStatus, container, dockerContainer, probe.Failure)
+						kl.readiness.set(dockerContainer.ID, ready == probe.Success)
+						containersToKeep[containerID] = &container
+						continue
+					}
+					kl.readiness.set(dockerContainer.ID, ready == probe.Success)
+					ref, ok := kl.getRef(containerID)
+					if !ok {
+						glog.Warningf("No ref for pod '%v' - '%v'", containerID, container.Name)
+					} else {
+						record.Eventf(ref, "unhealthy", "Liveness Probe Failed %v - %v", containerID, container.Name)
+					}
+					glog.Infof("pod %q container %q is unhealthy (probe result: %v). Container will be killed and re-created.", podFullName, container.Name, live)
 				} else {
-					kl.readiness.set(dockerContainer.ID, false)
-				}
-				if live == probe.Success {
-					containersToKeep[containerID] = empty{}
-					continue
+					createPodInfraContainer = true
+					delete(containersToKeep, podInfraContainerID)
+					// If we are to restart Infra Container then we move containersToKeep into containersToStart
+					// if RestartPolicy allows restarting failed containers. 
+					 if pod.Spec.RestartPolicy.Never == nil {
+						for _, v := range containersToKeep {
+							containersToStart[v] = empty{}
+						}
+						containersToKeep = make(map[dockertools.DockerID]*api.Container)
+					}
+					glog.Infof("pod %q container %q hash changed (%d vs %d). Pod will be killed and re-created.", podFullName, container.Name, hash, expectedHash)
 				}
-				ref, ok := kl.getRef(containerID)
-				if !ok {
-					glog.Warningf("No ref for pod '%v' - '%v'", containerID, container.Name)
-				} else {
-					record.Eventf(ref, "unhealthy", "Liveness Probe Failed %v - %v", containerID, container.Name)
+				// If RestartPolicy is Always or OnFailure we restart containers that were running before we
+				// killed them when restarting Infra Container.
+			} else {  // createPodInfraContainer == true
+				if pod.Spec.RestartPolicy.Never == nil {
```

Comment:
We can probably do this logic in one place, we do it in quite a few places today. Once we know we need to restart the podContainer we add all containers to the restart queue and check them one by one to remove them if they say restart never.